### DISCUSSION TITLE
Is the io_uring_sq thread in user mode or kernel mode, and is this a configurable option?

### DISCUSSION THREAD

**Discussion by wwwwxxxxhhhh** [axboe/liburing#1148] [Q&A] [answered]
The new version of io_uring_sq thread in library is in user mode, while the previous 0.7 version was in kernel mode. Is this a configurable option? Where is it configured? But the io_wQE_worker thread is still in kernel mode, right?

**Comment by axboe** [comment, marked as answer]
How the sq thread operates has nothing to do with the liburing version, it's strictly a kernel property. Some earlier kernels had kthreads, anything worth running (5.10-stable and newer) will have them be actual threads. Shared for both is that they only run in the kernel, they never exit to userspace. This is not something you can configure.

The sq thread and io-wq workers are the same kind of threads, again only depends on the kernel being used.

**Reply by wwwwxxxxhhhh** [reply]
Does it mean that in different kernel versions, SQ threads and worker threads may be in user mode or kernel mode?

**Reply by isilence** [reply]
It always runs in kernel as a kernel thread, there is no "user mode". Please clarify what you mean by "user mode" SQ threads.

axboe is talking about some differences of how it works internally, but it doesn't appear that's what you're asking.

**Reply by wwwwxxxxhhhh** [reply]
In the/proc/pid/stack file of the liburing0.7 version, it was found that the io_sq_thread thread contained the kthread keyword. However, in the/proc/pid/stack file of the liburing2.6 version, this keyword cannot be found. Therefore, do I determine if the io_sq_thread thread in the liburing2.6 version was created in user mode instead of kernel mode?
![image](https://github.com/axboe/liburing/assets/169776498/fb9b7dad-2d46-4c37-9c2c-fe256ce92db9)

**Reply by axboe** [reply]
I already explained this - it's not liburing specific, has nothing to do with liburing. Older kernels used kernel threads that assumed the identity of the user application, newer kernels use a regular kind of thread (not a kernel thread), but they only live in the kernel.

None of this is really interesting or pertinent to anything, it's internal details.

**Reply by wwwwxxxxhhhh** [reply]
Okay, I understand. Thank you very much for your reply.
